### ISSUE TITLE
LocalizedException in StoreRepository->get triggers InfiniteLoop (OutOfMemory) when currentStore cannot be calculated

### ISSUE THREAD

**Issue by mxpgmbh** [magento/magento2#37393]
### Preconditions (*)

1.  "Add Store Code to Urls" option has to be enabled 

### Steps to reproduce (*)

This is a bit tricky to reproduce with a minimal example. We noticed the problem when exporting Amasty GiftCards (https://amasty.com/gift-card-for-magento-2.html) with the Techdivision pixi exporter (https://docs.met.tdintern.de/pixi/3.x/index.html only in german though).\
**This is nonetheless a generic design problem, see below for a description**

1. Have "Add Store Code to Urls" set to true
2. Be in a custom route that doesn't use the store code URL in its path (e.g. /pixi/api/exportCatalog)
3. Set the current store to `null` via the StoreManager (`storeManager->setCurrentStore(null);`)
4. Load the store in this context (`$storeManager->getStore();`)

### Expected result (*)

1. An Exception with the text `The store that was requested wasn't found. Verify the store and try again.`, thrown by the StoreRepository->get function, because the route path is interpreted as store code (`pixi` in the above example).

### Actual result (*)

1. `Fatal error: Allowed memory size of 792723456 bytes exhausted (tried to allocate 20480 bytes) in /var/www/html/magento/vendor/magento/module-store/Model/StoreRepository.php on line 75`

The Problem is the usage of a `LocalizedException` in the `get` function:

```php
if ($store->getId() === null) {
    throw new NoSuchEntityException(
        __("The store that was requested wasn't found. Verify the store and try again.")
    );
}
```

Here the exception string should be translated. But to do that the current store has to be loaded to identify the correct language. Since it was set to null it tries to load the current store, which tries to load the incorrect store from the url path (`pixi` in this example), which results in this exception again, and so on, until the memory is exhausted.

Using a `LocalizedException` here is a design problem, because the translation cannot be done under such circumstances =>  Use another Exception with a non translated text.

**Comment by engcom-November** [comment]
Hello @mxpgmbh, 

Thank you for the report and collaboration!

Can you please elaborate the second step and it would be great if you could provide screenshot or example code for reference, this will help us to reproduce the issue more effectively.

Thank you.

**Comment by lc-skianet** [comment]
@mxpgmbh Hello, we are experiencing the same issue with the same module Amasty GiftCards and it happens when a Gift Card is added to the cart from the product page. It seems that the mini cart needs to be refreshed through this endpoint https://mymagento/customer/section/load/?sections=cart%2Cdirectory-data%2Cammessages%2Cmessages&force_new_section_timestamp=true&_=1700572840684 but calling this url results in an OOM exception, which is caused by the described steps 3 and 4. 

Removing the line `storeManager->setCurrentStore(null);` from the module fixes the problem, but we are not sure what side effects might arise. While debugging we also tracked down the issue to the `LocalizedException` in the `get` function inside the `/var/www/html/magento/vendor/magento/module-store/Model/StoreRepository.php` file at line 75. Did you find a fix or a workaround? Is the `storeManager->setCurrentStore(null);` needed for a specific reason?

Thank you!

**Comment by engcom-November** [comment]
Hello @mxpgmbh,

We tried to reproduce this issue on 2.4-develop, but in our case it's not reproducible.
Created a custom module and made the current store to null while `add store code to urls` is enabled, but the issue was not reproduced.
Please take a look at the screenshot below:
<img width="1465" alt="image" src="https://github.com/magento/magento2/assets/60198231/95d786e1-7be6-4088-aadb-9edf5808b846">
Here store code is enabled, and we are in a route which is not using store code, but 
Also in the url options in the admin area it's mentioned that "When using Store Code in URLs, in some cases system may not work properly if URLs without Store Codes are specified in the third-party services."
<img width="1051" alt="image" src="https://github.com/magento/magento2/assets/60198231/2701a97e-cb3f-415c-aa49-0de8d92727cc">

Please find the attached module and let us know if we are missing anything.
[MageVendor.zip](https://github.com/magento/magento2/files/13989131/MageVendor.zip)

Thank you.

**Comment by engcom-November** [comment]
Hello @mxpgmbh,

As there is no activity on this issue for a long time, we believe the issue has been resolved, hence closing this issue.
Feel free to raise a new issue or reopen this if you need more assistance.

Thank you.

**Comment by mpchadwick** [comment]
We are experiencing this

**Comment by hostep** [comment]
Sounds similar to https://github.com/magento/magento2/issues/39742, so possibly fixed by https://github.com/magento/magento2/commit/87bc610555e48d9c1e20c3e8dc4cd2128462aa7f (which got included in Magento 2.4.8)
